subprojects/core/src/main/java/org/gradle/internal/service/scopes/GradleUserHomeScopeServices.java

```diff
@@ -188,8 +193,36 @@ public void beforeComplete() {
         return fileSystemMirror;
     }
 
-    FileSystemSnapshotter createFileSystemSnapshotter(FileHasher hasher, StringInterner stringInterner, Stat stat, FileSystemMirror fileSystemMirror) {
-        return new DefaultFileSystemSnapshotter(hasher, stringInterner, stat, fileSystemMirror, DirectoryScanner.getDefaultExcludes());
+    VirtualFileSystem createVirtualFileSystem(FileHasher hasher, StringInterner stringInterner, Stat stat, ListenerManager listenerManager) {
+        DefaultVirtualFileSystem vfs = new DefaultVirtualFileSystem(hasher, stringInterner, stat, DirectoryScanner.getDefaultExcludes());
+        listenerManager.addListener(new OutputChangeListener() {
+            @Override
+            public void beforeOutputChange() {
+                vfs.invalidateAll();
+            }
+
+            @Override
+            public void beforeOutputChange(Iterable<String> affectedOutputPaths) {
+                vfs.update(affectedOutputPaths, () -> {});
+            }
+        });
+        listenerManager.addListener(new RootBuildLifecycleListener() {
+            @Override
+            public void afterStart() {
+            }
+
+            @Override
+            public void beforeComplete() {
+                vfs.invalidateAll();
+            }
+        });
+        return vfs;
+    }
+
+    FileSystemSnapshotter createFileSystemSnapshotter(FileHasher hasher, StringInterner stringInterner, Stat stat, FileSystemMirror fileSystemMirror, VirtualFileSystem vfs) {
+        return VFS_ENABLED
+            ? new VfsFileSystemSnapshotter(vfs, stringInterner, hasher)
+            : new DefaultFileSystemSnapshotter(hasher, stringInterner, stat, fileSystemMirror, DirectoryScanner.getDefaultExcludes());
     }
```

Comment:
Why again do we have two instances of `FileSystemSnapshotter`?

Reply:
We have to places where we save the state: in the Gradle user home and in the project home (= build session scope). We have now two virtual file systems, and a router between them.